dump/mapping/bufr_adpupa_prepbufr.py

```diff
@@ -0,0 +1,297 @@
+
+        if log_method == logger.info and level.upper() not in log_methods:
+            # Log a warning if the level is invalid
+            logger.warning(f'log level = {level}: not a valid level --> set to INFO')
+
+        # Call the logging method
+        log_method(message)
+
+
+def Compute_dateTime(cycleTimeSinceEpoch, hrdr):
+
+    int64_fill_value = np.int64(0)
+    dateTime = np.zeros(hrdr.shape, dtype=np.int64)
+    for i in range(len(dateTime)):
+        if ma.is_masked(hrdr[i]):
+            continue
+        else:
+            dateTime[i] = np.int64(hrdr[i]*3600) + cycleTimeSinceEpoch
+
+    dateTime = ma.array(dateTime)
+    dateTime = ma.masked_values(dateTime, int64_fill_value)
+
+    return dateTime
+
+
+def Mask_typ_for_var(typ, var):
+
+    typ_var = copy.deepcopy(typ)
+    for i in range(len(typ_var)):
+        if ma.is_masked(var[i]):
+            typ_var[i] = typ_var.fill_value
+
+    return typ_var
+
+
+def _make_description(mapping_path, cycle_time, update=False):
+    description = bufr.encoders.Description(mapping_path)
+
+    ReferenceTime = np.int64(calendar.timegm(time.strptime(str(int(cycle_time)), '%Y%m%d%H')))
+
+    if update:
+        # Define the variables to be added in a list of dictionaries
+        variables = [
+            #{
+            #    'name': 'ObsType/stationPressure',
+            #    'source': 'variables/stationPressureObservationType',
+            #    'units': '1',
+            #    'longName': 'Observation Type',
+            #},
+        ]
+
+        # Loop through each variable and add it to the description
+        for var in variables:
+            description.add_variable(
+                name=var['name'],
+                source=var['source'],
+                units=var['units'],
+                longName=var['longName']
+            )
+
+        description.add_global(name='datetimeReference', value=str(ReferenceTime))
+
+    return description
+
+
+def _make_obs(comm, input_path, mapping_path, cycle_time):
+    """
+    Create the ioda adpupa prepbufr observations:
+    - reads values
+
+    Parameters
+    ----------
+    comm: object
+            The communicator object (e.g., MPI)
+    input_path: str
+            The input bufr file
+    mapping_path: str
+            The input bufr2ioda mapping file
+    cycle_time: str
+            The cycle in YYYYMMDDHH format
+    """
+
+    # Get container from mapping file first
+    logging(comm, 'INFO', 'Get container from bufr')
+    container = bufr.Parser(input_path, mapping_path).parse(comm)
+
+    logging(comm, 'DEBUG', f'container list (original): {container.list()}')
+    logging(comm, 'DEBUG', f'prepbufrDataLevelCategory')
+    cat = container.get('variables/prepbufrDataLevelCategory')
+
+    logging(comm, 'DEBUG', f'Do DateTime calculation')
+    hrdr = container.get('variables/obsTimeMinusCycleTime')
+    hrdr_paths = container.get_paths('variables/obsTimeMinusCycleTime')
```

Comment:
you never use this result... same for other paths

Reply:
@rmclaren done, thanks!